extensions/amp-sidebar/1.0/sidebar.js

```diff
@@ -146,18 +147,49 @@ function SidebarWithRef(
   useLayoutEffect(() => {
     const sidebarElement = sidebarRef.current;
     const backdropElement = backdropRef.current;
-    if (!sidebarElement || !backdropElement) {
-      return;
-    }
-
-    if (!side) {
+    if (!side || !sidebarElement || !backdropElement) {
       return;
     }
 
     let sidebarAnimation;
     let backdropAnimation;
+    let keydownCallback;
+
+    const postInvisibleAnim = () => {
+      if (onAfterCloseRef.current) {
+        onAfterCloseRef.current();
+      }
+      sidebarAnimation = null;
+      backdropAnimation = null;
+      setMounted(false);
+    };
+    const documentElement =
+      sidebarElement.ownerDocument &&
+      sidebarElement.ownerDocument.documentElement;
+
     // "Make Visible" Animation
     if (opened) {
+      if (documentElement) {
+        keydownCallback = (event) => {
+          // Close sidebar on ESC.
+          if (event.key == Keys.ESCAPE) {
+            if (sidebarAnimation) {
+              sidebarAnimation.onfinish = () => {
+                postInvisibleAnim();
+                setOpened(false);
+              };
+              sidebarAnimation.reverse();
+            }
+            if (backdropAnimation) {
+              backdropAnimation.reverse();
+            }
+            event.preventDefault();
+            documentElement.removeEventListener('keydown', keydownCallback);
```

Comment:
Remove the callback once the `esc` key is pressed.  This is mostly useful when the sidebar is already fully opened.  Then the user clicks the `esc` key and it begins closing.

The callback actually reverses the existing animation, so if the sidebar is starting to close, and the `esc` key is clicked, then it will open back up.  That's why we remove the `eventListener` here.  (After the first `esc` press, it cannot be pressed again.)

~~~~

In the case where the user closes the sidebar normally (w/o `esc` key), the `eventListener` is also not available, since it's removed in the cleanup of the last `useLayoutEffect` so in this case, user also cannot accidentally reverse the animation.

Reply:
I think this complication also comes from combining the effects. All we need is a separate effect that calls `close()` API (or the version of it) on <kbd>Esc</kbd> key.

Reply:
Updated logic into a new `useEffect`.